iterative/gcp/provider.go

```diff
@@ -324,7 +324,27 @@ func getProjectService() (string, *gcp_compute.Service, error) {
 	}
 
 	if credentials.ProjectID == "" {
-		return "", nil, errors.New("Couldn't extract the project identifier from the given credentials!")
+		// 	Coerce Credentials to handle GCP OIDC auth
+		//	Common ProjectID ENVs:
+		//		https://github.com/google-github-actions/auth/blob/b05f71482f54380997bcc43a29ef5007de7789b1/src/main.ts#L187-L191
+		//		https://github.com/hashicorp/terraform-provider-google/blob/d6734812e2c6a679334dcb46932f4b92729fa98c/google/provider.go#L64-L73
+		coercedProjectID := utils.MultiEnvLoadFirst([]string{
+			"CLOUDSDK_CORE_PROJECT",
+			"CLOUDSDK_PROJECT",
+			"GCLOUD_PROJECT",
+			"GCP_PROJECT",
+			"GOOGLE_CLOUD_PROJECT",
+			"GOOGLE_PROJECT",
+		})
+		if coercedProjectID == "" {
+			// last effort to load
+			fromCredentialsID, err := utils.GCPCoerceOIDCCredentials(credentialsData)
```

Comment:
```suggestion
			fromCredentialsID, err := utils.GCPCoerceOIDCCredentials(credentials.JSON)
```

Same effect, cleaner data flow? 🤔

Reply:
I didn't test that or look to see in the `google.CredentialsFromJSON` does anything fishy where that might not be the same.

Reply:
[Luckily](https://cs.opensource.google/go/x/oauth2/+/97805856:google/default.go;l=196) it [doesn't](https://cs.opensource.google/go/x/oauth2/+/97805856:google/default.go;l=180) 😅

Reply:
regardless it still needs to exist for: https://github.com/iterative/terraform-provider-iterative/blob/2dd6c603529880ce3f942d44d916ff03f1dbce12/iterative/gcp/provider.go#L311-L312

Reply:
👍 and ill see if getting rid of it works

Reply:
```diff 
diff --git a/iterative/gcp/provider.go b/iterative/gcp/provider.go
index 3825dfd..87a85de 100644
--- a/iterative/gcp/provider.go
+++ b/iterative/gcp/provider.go
@@ -289,6 +289,14 @@ func ResourceMachineDelete(ctx context.Context, d *schema.ResourceData, m interf
        return nil
 }
 
+func LoadGCPCredentials() (*google.Credentials, error) {
+       if credentialsData := os.Getenv("GOOGLE_APPLICATION_CREDENTIALS_DATA"); credentialsData != "" {
+               return google.CredentialsFromJSON(oauth2.NoContext, []byte(credentialsData), gcp_compute.ComputeScope)
+       }
+
+       return google.FindDefaultCredentials(oauth2.NoContext, gcp_compute.ComputeScope)
+}
+
 func getServiceAccountData(saString string) (string, []string) {
        // ["SA email", "scopes=s1", "s2", ...]
        splitStr := strings.Split(saString, ",")
@@ -305,15 +313,7 @@ func getServiceAccountData(saString string) (string, []string) {
 }
 
 func getProjectService() (string, *gcp_compute.Service, error) {
-       var credentials *google.Credentials
-       var err error
-       credentialsData := []byte(utils.LoadGCPCredentials())
-       if len(credentialsData) > 0 {
-               credentials, err = google.CredentialsFromJSON(oauth2.NoContext, credentialsData, gcp_compute.ComputeScope)
-       } else {
-               credentials, err = google.FindDefaultCredentials(oauth2.NoContext, gcp_compute.ComputeScope)
-       }
-
+       credentials, err := LoadGCPCredentials()
        if err != nil {
                return "", nil, err
        }
diff --git a/iterative/resource_runner.go b/iterative/resource_runner.go
index 5a86449..adb183c 100644
--- a/iterative/resource_runner.go
+++ b/iterative/resource_runner.go
@@ -16,6 +16,7 @@ import (
 
        "gopkg.in/alessio/shellescape.v1"
 
+       "terraform-provider-iterative/iterative/gcp"
        "terraform-provider-iterative/iterative/utils"
 
        "github.com/hashicorp/terraform-plugin-sdk/v2/diag"
@@ -435,6 +436,11 @@ func provisionerCode(d *schema.ResourceData) (string, error) {
                return code, err
        }
 
+       var gcpCredentials string
+       if credentials, err := gcp.LoadGCPCredentials(); err == nil {
+               gcpCredentials = string(credentials.JSON)
+       }
+
        data := make(map[string]interface{})
        data["token"] = d.Get("token").(string)
        data["repo"] = d.Get("repo").(string)
@@ -455,7 +461,7 @@ func provisionerCode(d *schema.ResourceData) (string, error) {
        data["AZURE_CLIENT_SECRET"] = os.Getenv("AZURE_CLIENT_SECRET")
        data["AZURE_SUBSCRIPTION_ID"] = os.Getenv("AZURE_SUBSCRIPTION_ID")
        data["AZURE_TENANT_ID"] = os.Getenv("AZURE_TENANT_ID")
-       data["GOOGLE_APPLICATION_CREDENTIALS_DATA"] = utils.LoadGCPCredentials()
+       data["GOOGLE_APPLICATION_CREDENTIALS_DATA"] = gcpCredentials
        data["KUBERNETES_CONFIGURATION"] = os.Getenv("KUBERNETES_CONFIGURATION")
        data["container"] = isContainerAvailable(d.Get("cloud").(string))
        data["setup"] = strings.Replace(string(setup[:]), "#/bin/sh", "", 1)
diff --git a/iterative/utils/helpers.go b/iterative/utils/helpers.go
index 8797f25..bc83f0b 100644
--- a/iterative/utils/helpers.go
+++ b/iterative/utils/helpers.go
@@ -123,18 +123,6 @@ func GCPCoerceOIDCCredentials(rawCreds []byte) (string, error) {
        return projectID, nil
 }
 
-func LoadGCPCredentials() string {
-       credentialsData := os.Getenv("GOOGLE_APPLICATION_CREDENTIALS_DATA")
-       if len(credentialsData) == 0 {
-               credentialsPath := os.Getenv("GOOGLE_APPLICATION_CREDENTIALS")
-               if len(credentialsPath) > 0 {
-                       jsonData, _ := os.ReadFile(credentialsPath)
-                       credentialsData = string(jsonData)
-               }
-       }
-       return credentialsData
-}
-
 // Better way than copying?
 // https://github.com/hashicorp/terraform-provider-google/blob/8a362008bd4d36b6a882eb53455f87305e6dff52/google/service_scope.go#L5-L48
 func canonicalizeServiceScope(scope string) string {
```